src/utils/token_to_fiat_oracle.ts

```diff
@@ -0,0 +1,137 @@
+ * the browser platform by just removing node dependencies.
+ *
+ * https://github.com/miscavage/CoinGecko-API/issues/19
+ *
+ * When the report is taken and fixed, please:
+ * 1. Import the CoinGecko wrapper.
+ * 2. Define a CoinGecko client as private readonly at the
+ *   constructor arguments. So we can test it.
+ * 3. On each call to the API, replace the fetch for the actual
+ *   API call.
+ *
+ * Examples
+ *
+ * 1. import CoinGecko from 'coingecko-api'
+ * 2. constructor(private readonly client = new CoinGecko()) {}
+ * 3. const response = await this.client.ping() // for the ping() function
+ *    const response = await this.client.simple.price(params) // for getting the price of a currency
+ */
+
+export class TokenToFiatOracle implements FiatOracle {
+	private cachedPrices: TokenFiatRate[] = [];
+	private cacheTimestamp = 0;
+	private syncPromise?: Promise<void>;
+
+	/**
+	 * @param {FiatID[]} fiats The currency IDs to quote tokens in
+	 * @param {TokenID[]} tokens The token ID of the requested coin
+	 * @param {number} cacheLifespan Milliseconds that has to pass until the cache is invalid
+	 * @param {@link fetch} _fetch The standard fetch method. Defined here for testing propuses
+	 * @return {TokenToFiatOracle} A class representing the oracle for the selected currencies
+	 */
+	constructor(
+		private readonly fiats: FiatID[] = ['USD'],
+		private readonly tokens: TokenID[] = ['ARWEAVE'],
+		private readonly cacheLifespan = pollingIntervalMilliseconds,
+		private readonly _fetch = fetch
+	) {}
+
+	private get currentlyFetchingPrice(): boolean {
+		return !!this.syncPromise;
+	}
+
+	private get tokensCSV(): string {
+		return this.tokens.map(tokenIdToThirdParty).join(',');
+	}
+
+	private get fiatCSV(): string {
+		return this.fiats.map(fiatIdToThirdParty).join(',');
+	}
+
+	private get shouldRefreshCacheData(): boolean {
+		const currentTimestamp = Date.now();
+		const deltaTimestamp = currentTimestamp - this.cacheTimestamp;
+		return deltaTimestamp >= this.cacheLifespan;
+	}
+
+	/**
+	 * @param {FiatID} fiat The fiat currency ID
+	 * @param {TokenID} token The token ID
+	 * @throws {@link Error} If no such pair cached
+	 * @returns {Promise<TokenToFiatPrice>} The cached price value
+	 */
+	public async getPriceForFiatTokenPair(pair: TokenFiatPair): Promise<TokenFiatRate> {
+		await this.checkCache();
+		const pricePair = this.cachedPrices.find(cachePairFilterFactory(pair));
+		if (!pricePair) {
+			throw new Error(`No such pair (${pair.fiat}, ${pair.token})`);
+		}
+		return pricePair;
+	}
+
+	private async checkCache(): Promise<void> {
+		if (this.currentlyFetchingPrice) {
+			return this.syncPromise;
+		}
+		if (this.shouldRefreshCacheData) {
+			this.syncPromise = this.fetchPrices();
+			await this.syncPromise;
+			delete this.syncPromise;
+		}
+	}
+
+	private async fetchPrices(): Promise<void> {
+		const queryUrl = this.getQueryRequestUrl();
+		const fetchResponse = await this._fetch(queryUrl);
```

Comment:
For some reason the browser errors out when assigning fetch in this way:

```
Failed to use fiat oracle: TypeError: Failed to execute 'fetch' on 'Window': Illegal invocation
    at TokenToFiatOracle.fetchPrices (token_to_fiat_oracle.ts:99)
    at TokenToFiatOracle.checkCache (token_to_fiat_oracle.ts:91)
    at TokenToFiatOracle.getPriceForFiatTokenPair (token_to_fiat_oracle.ts:78)
    at getFiatToken (useFiatOracle.ts:10)
    at useFiatOracle.ts:25
 ```
 
 Assigning it directly works as intended in the browser, but does not solve the stubbing it out for testing issue.

Reply:
"Illegal invocation" lol :o 
Well, here I applied a change @arielmelendez suggested: To split that class into two:
- **TokenToFiatOracle** the actual oracle
- **CachingTokenToFiatOracle** the class I was writing, now only managing the cache

So now, instead of taking a fetch() for testing, the constructor will now take a TokenToFiatOracle which can easily be stubbed. Stubbing fetch was giving a lot of troubles

Reply:
Here it is the change: 64dde013c980c09bcf12acf2c6e1f62e723481c9